src/operator/distinct.rs

```diff
@@ -281,177 +268,248 @@ where
+    /// `t2` such that `not (t2 <= t1)` at which we've encountered `(k,v)`
+    /// and record `(k,v)` in `keys_of_interest[t']`.  When evaluating the
+    /// operator at time `t'` we simply scan all tuples in
+    /// `keys_of_interest[t2]` for candidates.
+    fn eval_keyval(
         &mut self,
-        trace_cursor: &mut T::Cursor<'s>,
-        _trace: &'s T,
-        value: &Z::Key,
-        weight: Z::R,
-        builder: &mut Z::Builder,
+        key: &Z::Key,
+        val: &Z::Val,
+        trace_cursor: &mut T::Cursor<'_>,
+        output: &mut Z::Builder,
     ) {
-        //eprintln!("value: {:?}, weight: {:?}", value, weight);
-        trace_cursor.seek_key(value);
-
-        if trace_cursor.key_valid() && trace_cursor.key() == value {
-            let mut w1: Z::R = HasZero::zero();
-            let mut w2: Z::R = HasZero::zero();
-            let mut w3: Z::R = HasZero::zero();
-            let mut next_ts: Option<T::Time> = None;
-            trace_cursor.map_times(|t, w| {
-                if !t.epoch() {
-                    if t.inner() < self.time {
-                        w1.add_assign_by_ref(w);
-                    } else if t.inner() == self.time {
-                        w2.add_assign_by_ref(w);
-                    } else if next_ts.is_none() || t < next_ts.as_ref().unwrap() {
-                        next_ts = Some(t.clone());
+        // println!("eval_keyval: {key:?}, {val:?}");
+        if trace_cursor.get_key() == Some(key) {
+            trace_cursor.seek_val(val);
+
+            if trace_cursor.get_val() == Some(val) {
+                // The nearest future timestamp when we need to update this
+                // key/value pair.
+                let mut time_of_interest = None;
+
+                // Reset all counters to 0.
+                self.clear_distinct_vals();
+
+                trace_cursor.map_times(|trace_ts, weight| {
+                    // Update weights in `distinct_vals`.
+                    for (ts, total_weight) in self.distinct_vals.iter_mut() {
+                        if let Some(ts) = ts {
+                            if trace_ts.less_equal(ts) {
+                                *total_weight += weight.clone();
+                            }
+                        }
+                    }
+                    // Timestamp in the future - update `time_of_interest`.
+                    if !trace_ts.less_equal(&self.time) {
+                        time_of_interest = match time_of_interest.take() {
+                            None => Some(self.time.join(trace_ts)),
+                            Some(time_of_interest) => {
+                                Some(min(time_of_interest, self.time.join(trace_ts)))
+                            }
+                        }
+                    }
+                });
+
+                // Compute `dist` for each entry in `distinct_vals`.
+                for (_time, weight) in self.distinct_vals.iter_mut() {
```

Comment:
```suggestion
                for weight in self.distinct_vals.values_mut() {
```

Reply:
This doesn't work on `Vec`'s

Reply:
Ah that's my bad, I thought it was a map